File: wayang-commons/wayang-basic/src/main/java/org/apache/wayang/basic/data/Row.scala
package org.apache.wayang.basic.data

/**
 * Represents a single row of data with an associated schema.
 * @param schema The metadata describing the structure of the fields.
 */
case class Row(fields: List[Any], schema: Schema = null) {

  val size: Int = fields.size

  def getFields: List[Any] = fields

  def get(index: Int): Any = fields(index)

  def getSchema: Schema = schema

}

/**
 * Of course a Schema is more complicated than the following Map, what follows is useful to
 * communicate the core of the Schema.
 */
case class Schema (columnNames: Map[String, Class[_]])

File: wayang-commons/wayang-basic/src/main/java/org/apache/wayang/basic/data/Row.java
package org.apache.wayang.basic.data;

import java.util.List;

/**
 * Represents a single row of data with an associated schema.
 * @param schema The metadata describing the structure of the fields.
 */
public record Row(List<Object> fields, Schema schema) {

    public int size() {
        return fields != null ? fields.size() : 0;
    }

    public Object getAttribute(String attribute) {
        int indexOfAttribute = schema.getIndexOfAttribute(attribute);
        if (indexOfAttribute >= 0) {
            return this.getAttributeAtIndex(indexOfAttribute);
        }
        return null;
    }

    private Object getAttributeAtIndex(int index) {
        return fields != null ? fields.get(index) : null;
    }
}